Move from webpack to esbuild bundler

martinpitt opened the pull request:
This speeds up the bundle build considerably, from about 26 s with    
webpack to about 6s with esbuild.    
    
Drop babel. esbuild understands JSX natively, and we don't need the    
"transpile to ES5" any more -- all recent browsers understand ES6.    
    
esbuild also creates LEGAL.txt files in development mode (unlike webpack
terser), so we can drop this hack as well.    

Co-Authored-By: Martin Pitt <mpitt@redhat.com> 

TODO:   
- [x] Builds on top of PR #1209 
- [x] move plugins to cockpit.git and if possible, make them compatible with both webpack and esbuild: https://github.com/cockpit-project/cockpit/pull/18372 → PR #1211 
- [x] Bundle CSS sizes are much bigger: https://github.com/evanw/esbuild/issues/2933; devel: 1.3 → 1.9 MB, prod: 97 → 146 kB (when gzipped)
- [x] Bundle JS size are also a bit bigger: 429 → 473 kB in production with gzip
- [x] Update to new pf-react-table which fixes the bundle size: PR #1214; rebase afterwards
- [x] fail on eslint/stylelint warnings: https://github.com/evanw/esbuild/issues/2939; fixed
- [x] (not a blocker, but ugly) sass plugin pulls in older/duplicate esbuild compiler: https://github.com/glromeo/esbuild-sass-plugin/issues/122
- [x] give some "building"/progress feedback in watch mode
- [x] bring back file gzipping for `NODE_ENV=production`
- [x] Move our custom plugins to cockpit.git https://github.com/cockpit-project/cockpit/pull/18404 and import from there
- [x] Decide what to do about the `esbuild` binary. Bundling in node_modules works for machines/podman/cockpit where we just rpm the pre-built dist/. But not for starter-kit and certificates where rpm build rebuilds the webpack (which is the cleaner approach); so let's restrict that to these projects for now.
- [x] Wait with landing until RHEL translation updates are done, and new point release: https://bugzilla.redhat.com/show_bug.cgi?id=2139647
- [x] clean up dist/ in between runs

All credit for this goes to @KKoukiou . I am mostly sending this PR to give it a trial run on CI, and collect a TODO list

## Development: Cockpit now supports the esbuild bundler

[esbuild](https://esbuild.github.io/) is a much faster bundler alternative for [webpack](https://webpack.js.org/). It also  natively understands [JSX](https://reactjs.org/docs/introducing-jsx.html), so that it is not necessary any more to use [babel](https://babeljs.io/). Cockpit's shared components and bundler plugins in [pkg/lib/](https://github.com/cockpit-project/cockpit/tree/main/pkg/lib) now support both eslint and webpack, so that projects can pick either. [cockpit-podman](https://github.com/cockpit-project/cockpit-podman) moved to esbuild in this commit **TODO**: add link once this lands. You may choose to do the same for your project.

martinpitt commented:
@KKoukiou : I updated `esbuild-rsync-plugin.js` here to be usable with webpack *and* esbuild. I tested the unchanged file with both bundlers, with and without `$RSYNC`. Next step is to do the same for the po plugin, then update both of them in cockpit, then we can use them from the shared code import. WDYT?

garrett commented:
Note: Switching from our breakpoint SCSS import dance to rewriting the CSS breakpoints will help detangle our SCSS and will probably make it smaller in the resulting size too. And it'll also fix our breakpoints further (as it's still broken in some places). So it'd probably help here too.

Details @ 
- https://github.com/cockpit-project/cockpit/issues/17629

martinpitt commented:
@KKoukiou : So AFAICS: Land #1214,  rebase/rebuild this, add plugins to cockpit pkg/lib/, pull this in here, and collect trophy, right? :trophy:

martinpitt commented:
@KKoukiou : I added some more cleanups, reworked the commit message, and rebased to the cockpit build which includes the  plugins. From my POV this is ready now, it's just blocked on doing a new point release first for RHEL translations. Anything else from your side still?

martinpitt commented:
Unblocked, this is ready to land. @mvollmer , @marusak , @jelly : Please have a look as well -- any questions/concerns?

Thanks @KKoukiou for figuring out all the hard bits here!

jelly commented:
The analyzer is cool:

![image](https://user-images.githubusercontent.com/67428/221877024-6042841c-2a6f-411c-96fa-01e762df4a66.png)

jelly commented:
I found one minor bug when I run `NODE_ENV=production npm run build` twice.

```
[jelle@t14s][~/projects/cockpit-podman]%NODE_ENV=production npm run build > metafile.json
node:internal/errors:867
  const err = new Error(message);
              ^

Error: Command failed: gzip -9 dist/po.ko.js
gzip: dist/po.ko.js.gz already exists;  not overwritten

    at ChildProcess.exithandler (node:child_process:419:12)
    at ChildProcess.emit (node:events:512:28)
    at maybeClose (node:internal/child_process:1098:16)
    at Socket.<anonymous> (node:internal/child_process:456:11)
    at Socket.emit (node:events:512:28)
    at Pipe.<anonymous> (node:net:316:12) {
  code: 2,
  killed: false,
  signal: null,
  cmd: 'gzip -9 dist/po.ko.js',
  stdout: '',
  stderr: 'gzip: dist/po.ko.js.gz already exists;\tnot overwritten\n'
}
```

martinpitt commented on build.js:
```diff
@@ -0,0 +1,72 @@
+    target: ['es2020'],
+    plugins: [
+        ...stylelint ? [stylelintPlugin()] : [],
+        ...process.env.ESLINT !== '0' ? [eslintPlugin] : [],
```
You are right, fixed.

jelly commented:
Now with a production build!

![image](https://user-images.githubusercontent.com/67428/221886574-4718fa13-adb8-42cc-b22e-34192ff24756.png)

martinpitt commented:
@jelly 's issue is actually much worse.. It also happens with `NODE_ENV=production npm run watch` on the second build.

These are both a symptom of a deeper problem: The dist/ directory does not get cleaned between the builds. Do this:

    rm -rf dist; NODE_ENV=production npm run build; npm run build

Now dist/ has *both* the compressed (production) and uncompressed (devel) files. This has already annoyed me a lot with webpack, that does the same. So the use case is to start with a prod build (from e.g. `make vm` or `make check`), then wanting to fix something, running `make watch` and wondering why the changes don't apply -- it's because cockpit-ws prefers serving .gz files.

So we need to clean up old files, or possibly all files, in dist/ on each round. [webpack has a `clean: true` option](https://webpack.js.org/guides/output-management/#cleaning-up-the-dist-folder), we should check if esbuild has something similar, and if not, build it somehow.

martinpitt commented:
https://esbuild.github.io/api/#outdir documents this behaviour, and basically says "do it yourself". So I added a trivial `cleanPlugin`, all of the cases above work now (building twice, switching between prod and dev, watch mode). @KKoukiou , @jelly , PTAL?

martinpitt commented:
Wah, node_modules/ changes every two hours, rebuilt for the 128th time.

martinpitt commented:
After some discussion with @KKoukiou , I cleaned up how we handle the linters. Pushing with an [extra committ](https://github.com/cockpit-project/cockpit-podman/pull/1210/commits/0b1b5d9a7d381204488a0c7addcb35a928da4ad5) which should **fail** in CI. After this proves to works as intended, I'll take it out again.

martinpitt commented on build.js:
```diff
@@ -0,0 +1,87 @@
+const nodePaths = ['pkg/lib'];
+const outdir = 'dist';
+
+// always start with a fresh dist/ directory, to change between development and production, or clean up gzipped files
+const cleanPlugin = {
```
I suppose this should also eventually move to cockpit's pkg/lib/, but at this point my gut feeling is that we can land this PR, and do that separately. @KKoukiou is already working on esbuild-ifying cockpit.git, and she will most certainly add the clean plugin there as well.
